File: ruby/easy/design_compressed_string_iterator.rb
class StringIterator

=begin
    :type compressed_string: String
=end
    def initialize(compressed_string)
        @compressed_string = compressed_string
        @i = @count = 0
    end


=begin
    :rtype: Character
=end
    def next()
        return " " unless has_next()
        if(@count <= 0)
            @char = @compressed_string[@i]
            @i += 1
            @count = get_count()
        end
        @count -= 1
        return @char
    end


=begin
    :rtype: Boolean
=end
    def has_next()
        return @i < @compressed_string.length || @count > 0
    end

    private
    def get_count()
        count = 0
        while @i < @compressed_string.length
            c = @compressed_string[@i]
            break unless is_numeric(c)
            count = (count * 10) + get_value(c)
            @i += 1
        end
        return count
    end
    
    def is_numeric(c)
        return c >= '0' && c <= '9'
    end
    
    def get_value(c)
        return c.ord - '0'.ord
    end
end

# Your StringIterator object will be instantiated and called as such:
# obj = StringIterator.new(compressed_string)
# param_1 = obj.next()
# param_2 = obj.has_next()

File: javascript/easy/designCompressedStringIterator.js
/**
 * @param {string} compressedString
 */
var StringIterator = function(compressedString) {
    this.compressedString = compressedString;
    this.i = 0;
    this.count = 0;
    this.char = 0;
};

/**
 * @return {character}
 */
StringIterator.prototype.next = function() {
    if(!this.hasNext()) return " ";
    let i = this.i, compressedString = this.compressedString;
    if(this.count <= 0) {
        this.char = compressedString[i];
        i++;
        let count = 0;
        while(i < compressedString.length) {
            let c = compressedString[i];
            if(!StringIterator.isNumeric(c)) {
                break;
            }
            count = (count * 10) + StringIterator.getValue(c);
            i++;
        }
        this.count = count;
        this.i = i;
    }
    
    this.count--;
    return this.char;
};

/**
 * @return {boolean}
 */
StringIterator.prototype.hasNext = function() {
    return this.i < this.compressedString.length || this.count > 0;
};

StringIterator.isNumeric = function(c) {
    let code = c.charCodeAt(0);
    return code >= '0'.charCodeAt(0) && code <= '9'.charCodeAt(0);
};

StringIterator.getValue = function(c) {
    return c.charCodeAt(0) - '0'.charCodeAt(0);    
};

/** 
 * Your StringIterator object will be instantiated and called as such:
 * var obj = Object.create(StringIterator).createNew(compressedString)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
